### PR TITLE
REF: Consolidate validation of dictionary argument in agg/transform

### PR DESCRIPTION
**Pull request by rhshadrach**
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

Currently there are no tests for dict-like argument with agg when a DataFrame is missing a column (edit: in the apply tests, there are tests in groupby, resample, and window). This PR changes the error slightly from

> f"Column '{k}' does not exist!"

to

> f"Column(s) {cols} do not exist"

in order to make it consistent (and I think more helpful, in the case of multiple columns) with the error message from transform.

### CODE REVIEW

**Comment by rhshadrach** [comment]
@jreback - added test_invalid_arg to start reorganizing the apply tests to something with a better layout. If this seems like a good idea, will move other relevant tests here in a followup.

**Comment by jreback** [comment]
great, this is a very small api change (e.g. now get SpecificationError rather than KeyError in some cases), can you add a whatsnew note. ping on green.

**Comment by rhshadrach** [comment]
@jreback Ah - thanks for catching this. Before adding the note - is it clear that we want a SpecificationError here and not KeyError? Looking throughout pandas.core, I only see SpecificationError used in three places:

 - groupby.generic._aggregate_multiple_funcs
 - aggregation.reconstruct_func  - only used by DataFrame.agg and DataFrameGroupBy.agg
 - apply for transform (and with this PR, agg too)

In each case, I believe either KeyError or ValueError is also appropriate. It seems to me maybe changing to KeyError for transform here would be better, and removing SpecificationError altogether in the future. This would also having the benefit of a smaller API change (transform is only used by Series/DataFrame here, whereas agg is used by GroupBy/Resampler/Rolling).

File: pandas/tests/apply/test_invalid_arg.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+
+@pytest.mark.parametrize("box", [DataFrame, Series])
+@pytest.mark.parametrize("method", ["apply", "agg", "transform"])
+@pytest.mark.parametrize("func", [{"A": {"B": "sum"}}, {"A": {"B": ["sum"]}}])
+def test_nested_renamer(box, method, func):
+    # GH 35964
+    obj = box({"A": [1]})
+    match = "nested renamer is not supported"
+    with pytest.raises(SpecificationError, match=match):
```
Comment (jreback) [inline pandas/tests/apply/test_invalid_arg.py:20]: this error is good

File: pandas/tests/apply/test_invalid_arg.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        getattr(obj, method)(func)
+
+
+@pytest.mark.parametrize("method", ["apply", "agg", "transform"])
+@pytest.mark.parametrize("func", [{"B": "sum"}, {"B": ["sum"]}])
+def test_missing_column(method, func):
+    # GH 40004
+    obj = DataFrame({"A": [1]})
+    match = re.escape("Column(s) ['B'] do not exist")
+    with pytest.raises(SpecificationError, match=match):
```
Comment (jreback) [inline pandas/tests/apply/test_invalid_arg.py:30]: would be ok here with a KeyError (which i think is the same as now)

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
very nice!

**Pull request merged**